server/controllers/chartsController/index.js

```diff
@@ -1,80 +1,156 @@
+        const { color, label, value, targetValues } = fieldLabelValueMap
+        const targetValue = targetValues[study]
+        const siteName = SITE_NAMES[study] || study
+        const dataKey = `${siteName}-${label}-${targetValue}`
+        const totalsDataKey = `${TOTALS_STUDY}-${label}`
+        const isVariableValueEmpty = value === EMPTY_VALUE
+        const shouldCountSubject = isVariableValueEmpty
+          ? subjectDayData.every((day) => day[chart.variable] === value)
+          : subjectDayData.some((dayData) => dayData[chart.variable] == value)
 
-      labelMap.set(label, { name: label, color })
-      processData({ shouldCountSubject, dataMap, dataKey, totalsDataKey })
-      processTotals({
-        shouldCountSubject,
-        studyTotals,
-        siteName,
-        targetValue,
+        labelMap.set(label, { name: label, color })
+        processData({ shouldCountSubject, dataMap, dataKey, totalsDataKey })
+        processTotals({
+          shouldCountSubject,
+          studyTotals,
+          siteName,
+          targetValue,
+        })
       })
-    })
-  }
+    }
+
+    const dataBySite = postProcessData(dataMap, studyTotals)
 
-  const dataBySite = postProcessData(dataMap, studyTotals)
+    if (isAnyTargetIncluded(studyTotals)) {
+      labelMap.set(N_A, { name: N_A, color: '#808080' })
+    }
 
-  if (isAnyTargetIncluded(studyTotals)) {
-    labelMap.set(N_A, { name: N_A, color: '#808080' })
+    return {
+      chart,
+      dataBySite: Array.from(dataBySite.values()),
+      labels: Array.from(labelMap.values()),
+      studyTotals,
+      filters,
+    }
+  } else {
+    return {
+      chart,
+      dataBySite: [],
+      labels: Array.from(labelMap.values()),
+      studyTotals,
+      filters,
+    }
   }
+}
 
-  return {
-    chart,
-    dataBySite: Array.from(dataBySite.values()),
-    labels: Array.from(labelMap.values()),
-    studyTotals,
+const getAllSubjects = async ({
+  dataDb,
+  chart,
+  userAccess,
+  areFiltersRequested,
+  filters,
+}) => {
+  if (!areFiltersRequested) {
+    return await dataDb
+      .collection(collections.toc)
+      .find(
+        {
+          assessment: chart.assessment,
+          study: { $in: userAccess, $nin: STUDIES_TO_OMIT },
+        },
+        { projection: { collection: 1, study: 1, _id: 0 } }
+      )
+      .toArray()
+  } else {
+    const mongoFilterQuery = mongoQueryFromFilters(filters)
+    const filteredSubjects = []
+    const criteriaCursor = await dataDb.collection(collections.toc).find(
+      {
+        assessment: INCLUSION_EXCLUSION_CRITERIA_FORM,
+        study: { $in: userAccess, $nin: STUDIES_TO_OMIT },
+      },
+      {
+        projection: { collection: 1, study: 1, _id: 0, subject: 1 },
+      }
+    )
+
+    for await (let doc of criteriaCursor) {
+      const data = await dataDb
+        .collection(doc.collection)
+        .findOne({ $or: mongoFilterQuery })
+
+      if (data) filteredSubjects.push(data?.subjectid)
+    }
```

Comment:
if we call `toArray()` on the `criteriaCursor` request could we make it so we're not pushing onto the `filteredSubjects` array? Is it a bad idea to call `toArray` for performance reasons or otherwise? If we did change it to an array I think we would be able to change this to something like:

```ts
const filteredSubjects = await Promise.all(criteriaCursor.map(doc => {
  const data = await dataDb
    .collection(doc.collection)
    .findOne({ $or: mongoFilterQuery })

  return data?.subjectid
}))
```

Though we would then need to `filteredSubjects.filter(Boolean)` to get rid of empty items. 🤷

Reply:
There are a few reasons why I decided to go against using toArray.
1. toArray() needs to complete all of the lookups and then converts it into an array 
2. it is less performant 
3. We would be converting all of the results so we would still need to iterate the results and filter the ones that don't have data

In your example is that the Mongodb map? or the Array.map?

Reply:
ok I think your explanation makes sense and we should continue to do it this way. Thanks for the additional info!